### ISSUE TITLE
bootstrapping will fail if csr exists

### ISSUE THREAD

**Issue by alexeldeib** [krustlet/krustlet#634]
I believe this line should handle 409s: https://github.com/deislabs/krustlet/blob/d4d6417ff1fea4e988031a719b6b619bdd7df9af/crates/kubelet/src/bootstrapping/mod.rs#L98

It works if you already have a "full" (non-bootstrap) kubeconfig, because you won't hit this flow. But if you e.g. start krustlet for bootstrapping, let it create the csr, then rapidly restart it, it will fail with 409s on creating the CSR (because it already created it).

Instead, a 409 should probably trigger a check on whether the CSR is valid/approved, and then proceed accordingly, either by clearing/reissuing or using the current one.

looks like `csrs.create(...)` returns a `Result<K, E>` where K is the CSR and E is a [kube:Error](https://docs.rs/kube/0.57.0/kube/enum.Error.html) which has an `HttpError(Error)` variant, which itself wraps a generic [http:Error](https://docs.rs/http/0.2.4/http/struct.Error.html). So we could match on the initial error being an HttpError, then match on the HttpError being a 409. otherwise, if the result was an error, we should return it.

**Comment by thomastaylor312** [comment]
Yeah, we've known about this for a while, but didn't really document it. We'll add this to the milestone to fix before 1.0. If you already have an idea of fixing it, PRs are totally welcome! Just let us know if you are working on it.

**Comment by alexeldeib** [comment]
I'm not working on it yet, but I might after figuring out a few other rough edges :)

**Comment by technosophos** [comment]
I think `kube::Error::Api` is the variant you want to use. An HttpError means that the HTTP protocol itself broke in some way. I think the `409` would be reported on the `Api` variant.

That said... I am a little suspicious that a Kubelet should trust a CSR that it did not generate. Wouldn't this open a potential vector for an attack where a hacker could create a bad CSR and get Krustlet to generate a certificate based on that?

**Comment by technosophos** [comment]
I got at least the stub of a solution done here. I'm just not sure what should be done when a 409 occurs. Given the constraints on creating the CSR, it might actually be safe to do what I did there -- because only a CSR with the same properties should be created. But I am not sure about that at all, and I am not sure under which circumstances a CSR will persist after the cert has been signed.

https://github.com/krustlet/krustlet/pull/676
